File: backend/app/pipeline/score.py
"""Signature readiness score and category breakdown."""

from __future__ import annotations

from typing import Any

from app.pipeline.classify import ContractType


def compute_scores(
    contract_type: ContractType,
    clauses: list[dict[str, Any]],
    missing: list[dict],
) -> tuple[int, dict[str, int], str]:
    # Base 85, subtract for high/medium risks and missing items
    base = 85
    cat = {"money": 80, "termination": 80, "ip_privacy": 80, "disputes": 80, "flexibility": 80}

    for c in clauses:
        lvl = c.get("risk_level", "low")
        theme = c.get("theme", "general")
        penalty = 6 if lvl == "high" else 3 if lvl == "medium" else 0
        base -= penalty

        if theme == "payment" or theme == "deposit":
            cat["money"] -= penalty
        if theme == "termination" or theme == "renewal":
            cat["termination"] -= penalty
            cat["flexibility"] -= penalty // 2
        if theme in ("ip", "confidentiality"):
            cat["ip_privacy"] -= penalty
        if theme == "disputes":
            cat["disputes"] -= penalty

    for _ in missing:
        base -= 4
        cat["flexibility"] -= 3

    for k in cat:
        cat[k] = max(35, min(100, cat[k]))

    overall = max(25, min(100, base))

    if overall >= 75:
        rec = "mostly_standard"
    elif overall >= 55:
        rec = "worth_clarifying"
    elif overall >= 40:
        rec = "caution"
    else:
        rec = "strongly_consider_review"

    return overall, cat, rec

File: ios/CoParse/Engine/Score.swift
import Foundation

enum AnalysisScore {
    static func compute(
        contractType: ContractType,
        clauses: [ClauseCard],
        missing: [MissingProtection]
    ) -> (overall: Int, categoryScores: [String: Int], recommendation: String) {
        _ = contractType
        var base = 85
        var cat = ["money": 80, "termination": 80, "ip_privacy": 80, "disputes": 80, "flexibility": 80]

        for c in clauses {
            let penalty = c.riskLevel == "high" ? 6 : (c.riskLevel == "medium" ? 3 : 0)
            base -= penalty
            switch c.theme {
            case "payment", "deposit":
                cat["money", default: 80] -= penalty
            case "termination", "renewal":
                cat["termination", default: 80] -= penalty
                cat["flexibility", default: 80] -= penalty / 2
            case "ip", "confidentiality":
                cat["ip_privacy", default: 80] -= penalty
            case "disputes":
                cat["disputes", default: 80] -= penalty
            default:
                break
            }
        }

        for _ in missing {
            base -= 4
            cat["flexibility", default: 80] -= 3
        }

        for key in cat.keys {
            cat[key] = max(35, min(100, cat[key] ?? 80))
        }

        let overall = max(25, min(100, base))
        let rec: String
        if overall >= 75 { rec = "mostly_standard" }
        else if overall >= 55 { rec = "worth_clarifying" }
        else if overall >= 40 { rec = "caution" }
        else { rec = "strongly_consider_review" }

        return (overall, cat, rec)
    }
}
